boa_engine/src/module/mod.rs

```diff
@@ -0,0 +1,701 @@
+    ///
+    /// loader.insert(Path::new("main.mjs").to_path_buf(), module.clone());
+    ///
+    /// let promise = module.load_link_evaluate(context).unwrap();
+    ///
+    /// context.run_jobs();
+    ///
+    /// assert_eq!(promise.state().unwrap(), PromiseState::Fulfilled(JsValue::undefined()));
+    /// ```
+    #[allow(clippy::drop_copy)]
+    pub fn load_link_evaluate(&self, context: &mut Context<'_>) -> JsResult<JsPromise> {
+        let main_timer = Profiler::global().start_event("Module evaluation", "Main");
+
+        let promise = self
+            .load(context)
+            .then(
+                Some(
+                    FunctionObjectBuilder::new(
+                        context,
+                        NativeFunction::from_copy_closure_with_captures(
+                            |_, _, module, context| {
+                                module.link(context)?;
+                                Ok(JsValue::undefined())
+                            },
+                            self.clone(),
+                        ),
+                    )
+                    .build(),
+                ),
+                None,
+                context,
+            )?
+            .then(
+                Some(
+                    FunctionObjectBuilder::new(
+                        context,
+                        NativeFunction::from_copy_closure_with_captures(
+                            |_, _, module, context| Ok(module.evaluate(context).into()),
+                            self.clone(),
+                        ),
+                    )
+                    .build(),
+                ),
+                None,
+                context,
+            )?;
+
+        // The main_timer needs to be dropped before the Profiler is.
+        drop(main_timer);
+        Profiler::global().drop();
+
+        Ok(promise)
+    }
+
+    /// Abstract operation [`GetModuleNamespace ( module )`][spec].
+    ///
+    /// Gets the [**Module Namespace Object**][ns] that represents this module's exports.
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-getmodulenamespace
+    /// [ns]: https://tc39.es/ecma262/#sec-module-namespace-exotic-objects
+    pub fn namespace(&self, context: &mut Context<'_>) -> JsObject {
+        self.inner
+            .namespace
+            .borrow_mut()
+            .get_or_insert_with(|| {
+                let exported_names = self.get_exported_names(&mut Vec::default());
+
+                let unambiguous_names = exported_names
+                    .into_iter()
+                    .filter_map(|name| {
+                        self.resolve_export(name, &mut HashSet::default())
+                            .ok()
+                            .map(|_| name)
+                    })
+                    .collect();
+
+                ModuleNamespace::create(self.clone(), unambiguous_names, context)
+            })
+            .clone()
+    }
+}
+
+impl PartialEq for Module {
+    fn eq(&self, other: &Self) -> bool {
```

Comment:
Since these are `pub` functions I think they should be annotated with `#[inline]`